expression/builtin_cast.go

```diff
@@ -122,31 +122,31 @@ func (c *castAsIntFunctionClass) getFunction(ctx context.Context, args []Express
 		sig.setPbCode(tipb.ScalarFuncSig_CastIntAsInt)
 		return sig.setSelf(sig), nil
 	}
-	switch args[0].GetTypeClass() {
-	case types.ClassInt:
+	argTp := fieldTp2EvalTp(args[0].GetType())
+	switch argTp {
+	case tpInt:
 		sig = &builtinCastIntAsIntSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_CastIntAsInt)
-	case types.ClassReal:
+	case tpReal:
 		sig = &builtinCastRealAsIntSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_CastRealAsInt)
-	case types.ClassDecimal:
+	case tpDecimal:
 		sig = &builtinCastDecimalAsIntSig{bf}
 		sig.setPbCode(tipb.ScalarFuncSig_CastDecimalAsInt)
-	case types.ClassString:
-		tp := args[0].GetType().Tp
-		if types.IsTypeTime(tp) {
-			sig = &builtinCastTimeAsIntSig{bf}
-			sig.setPbCode(tipb.ScalarFuncSig_CastTimeAsInt)
-		} else if tp == mysql.TypeDuration {
-			sig = &builtinCastDurationAsIntSig{bf}
-			sig.setPbCode(tipb.ScalarFuncSig_CastDurationAsInt)
-		} else if tp == mysql.TypeJSON {
-			sig = &builtinCastJSONAsIntSig{bf}
-			sig.setPbCode(tipb.ScalarFuncSig_CastJsonAsInt)
-		} else {
-			sig = &builtinCastStringAsIntSig{bf}
-			sig.setPbCode(tipb.ScalarFuncSig_CastStringAsInt)
-		}
+	case tpDatetime, tpTimestamp:
+		sig = &builtinCastTimeAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastTimeAsInt)
+	case tpDuration:
+		sig = &builtinCastDurationAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastDurationAsInt)
+	case tpJSON:
+		sig = &builtinCastJSONAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastJsonAsInt)
+	case tpString:
+		sig = &builtinCastStringAsIntSig{bf}
+		sig.setPbCode(tipb.ScalarFuncSig_CastStringAsInt)
+	default:
+		panic("unsupported evalTp in castAsIntFunctionClass")
```

Comment:
Return an error is better.

Reply:
no, we can discover programmer's mistake through this panic.